### ISSUE TITLE
When you change hideDefaultLocaleInURL to true, access to the default language will appear 404

### ISSUE THREAD

**Issue by monkey-qiang1994** [mcamara/laravel-localization#656]
Hello, first of all thank you very much for your plugin, he is very good
I am having a problem now,When I change hideDefaultLocaleInURL to true, access to the default language will appear 404.
www.test.com/contact Can be accessed 
www.test.com/es/contact A 404 error has occurred and there is no redirect
I have already registered and added middleware
Waiting for your reply, thank you again for your plugin

**Comment by walidbagh** [comment]
Can confirm this happening on in my project !
Laravel 5.8

**Comment by monkey-qiang1994** [comment]
> Can confirm this happening on in my project !
> Laravel 5.8
> 
> EDIT: updated to latest version: 1.4.2 and same problem...

My Laravel is 5.5

**Comment by walidbagh** [comment]
I also noticed that despite having both : 
``` 
'useAcceptLanguageHeader' => true,   
'hideDefaultLocaleInURL' => true,
```
the language negotiation still happens so, when visiting the `/` route for the first time it changes to `/fr` instead of "app.locale" (en).

**Comment by iwasherefirst2** [comment]
> I also noticed that despite having both :
> 
> ```
> 'useAcceptLanguageHeader' => true,   
> 'hideDefaultLocaleInURL' => true,
> ```
> 
> the language negotiation still happens so, when visiting the `/` route for the first time it changes to `/fr` instead of "app.locale" (en).

This is the expected behavior of `"useAcceptLanguageHeader" => true` (see readme.md). If there would be no redirect when accessing`/` at the first time, then there would be no redirect ever. If that is what you want, then you may use `"useAcceptLanguageHeader" => false`

**Comment by walidbagh** [comment]
@iwasherefirst2 you got that right, my bad.
but i still can wrap my head over the 404 problem...

**Comment by monkey-qiang1994** [comment]
At the same time, I found that all my POST requests will also be wrong, showing blank or MethodNotAllowedHttpException No message... I don't know if there is any method I have not used? Has anyone encountered this problem?

**Comment by vagifrzada** [comment]
I'm also having same problem. When changing configuration to `'hideDefaultLocaleInURL' => true`, everything looks fine. But i need hide the default locale in url.

**Comment by satoved** [comment]
Just made "composer update" and got the same problem

**Comment by satoved** [comment]
Got back to 1.4.1 and the problem is gone. Something is off with the last release.

**Comment by hedii** [comment]
Having the same problem with 1.4.1 and 1.4.2 (i don't remember if I ever used 1.4.0)

````
'useAcceptLanguageHeader' => false,
'hideDefaultLocaleInURL' => false,
````

**Before 1.4.x**
/ redirects to /{locale}
/contact redirects to /{locale}/contact

**After 1.4.x**
/ is 404
/contact is 404

**Comment by DmitrySkibitsky** [comment]
`Updating mcamara/laravel-localization (1.4.1 => 1.4.2)`

**As a result, the error was not corrected.**

> 
> Such problem. It is observed only on a certain route. In the rest of the routes correctly changes localization. Changed package version to 1.3.2 but nothing helped.
> 
> Laravel 5.8.35
> 
> config in AppServiceProvider
> ```php
>  config(['laravellocalization.supportedLocales' => [
>             'ru' => ['name' => 'RU', 'script' => 'Cyrl'],
>             'ua' => ['name' => 'UA', 'script' => 'Cyrl']
>         ],
>             'laravellocalization.useAcceptLanguageHeader' => true,
>             'laravellocalization.hideDefaultLocaleInURL' => false
>         ]);
> ```
> 
> The screenshot showed the problematic route.
> ![DeepinScreenshot_выберите-область_20190912195505](https://user-images.githubusercontent.com/20288556/64804335-4b342f80-d597-11e9-99a8-ccef3b1d6046.png)
>

**Comment by Adzhantis** [comment]
reproduce

**Comment by iwasherefirst2** [comment]
The issue is caused by the PR #651 . See line 175 in `vendor/mcamara/laravel-localization/src/LaravelLocalization`. As a temporary hotfix you may uncomment this line and things should work.

The problem is caused, because inside of `routes.php` `setLocale` is called. If you hide default locale and try to access `/en/**` then with the newest release, `setLocale` will return `null` instead of `en`. Thus the route `/en/**` is unknown. Also the new behavior breaks the function description which states:

> Returns locale (if route has any) or null (if route does not have a locale)

Since the route `/en/**` has a locale, it should return `en`. 

I do not know much about the caching logic, but I think we should avoid change the logic of the `setLocale` method @czim and @mcamara

**Comment by elnurxf** [comment]
Same here. Reverted back to 1.4.0

**Comment by czim** [comment]
Sorry peeps, I think I got lost in the maze of when `setLocale` should return what. The 'fix' attempted to solve the problem that `setLocale` would return something other than `null`, when it should, when routes were cached. Now it appears it does the reverse.

That said, I'm personally not in favor of considering some parts of the code 'off limits' simply because we don't know what it should do when, exactly. I'd rather find out the fix that makes most sense semantically.

It does not help that route caching for multiple locales is a problem with an inherently hacky solution. I'm super open for new/better ideas and to try this again. I think new tests may also be needed to make sure there's no further regression like this.

**Comment by iwasherefirst2** [comment]
> It does not help that route caching for multiple locales is a problem with an inherently hacky solution. I'm super open for new/better ideas and to try this again. I think new tests may also be needed to make sure there's no further regression like this.

I will write a feature tests for `hideDefaultLocale` in the next days, since we have now the possibility to do  full feature tests (https://github.com/mcamara/laravel-localization/issues/161#issuecomment-526537412).

I can take a look into the caching functionality, but I won't be able to do so before 8th of october. I agree that we should look forward for a good universal solution, but when we change the `setLocale` behavior from its description, then we would have to create a new release 2.x for that in my opinion, because that would be a pretty big change, effecting all the middlewares and public API behavior.

**Comment by mcamara** [comment]
I can't make a PR now but if somebody does it and fixes it I'll merge it straightaway.

Thanks

**Comment by mcamara** [comment]
Should be fixed with the 1.4.3 version.

Thanks to @iwasherefirst2 for his help!
